### PR TITLE
[no squash] GL driver fixes & 10-bit texture format for post-processing

### PR DESCRIPTION
**Pull request by sfan5**
* fixes RTT creation with `ECF_A8R8G8B8` on GLES 3.0
* fixes incorrect handling of `GL_APPLE_texture_format_BGRA8888`
* adds 10-bit texture format we can use during post-processing to win some performance back (optional)
* fixes problems described in #15457
* some minor cleanups in driver and scene manager

## To do

This PR is ready.

### CODE REVIEW

**Comment by sfan5** [comment]
@grorp please test

File: src/defaultsettings.cpp [hunk 1]
```diff
@@ -565,7 +566,8 @@ void set_default_settings()
 	settings->setDefault("active_block_range", "2");
 	settings->setDefault("viewing_range", "50");
 	settings->setDefault("leaves_style", "simple");
-	settings->setDefault("enable_post_processing", "false");
+	settings->setDefault("enable_post_processing", "false"); // remove this?
```
Comment (grorp) [inline src/defaultsettings.cpp:569]: Removing "enable_post_processing = false" as an option would mean dropping support for OpenGL ES 2.0 without extensions afaict (https://github.com/minetest/minetest/issues/13519#issuecomment-2462938547)

(Also: the last time I tested, "enable_post_processing = false" still resulted in a performance improvement on my laptop)

File: irr/src/OpenGLES2/Driver.cpp [hunk 2]
```diff
@@ -86,10 +86,8 @@ void COpenGLES2Driver::initFeatures()
 		TextureFormats[ECF_R8G8B8] = {GL_RGB, GL_RGB, GL_UNSIGNED_BYTE};
 		TextureFormats[ECF_A8R8G8B8] = {GL_RGBA, GL_RGBA, GL_UNSIGNED_BYTE, CColorConverter::convert_A8R8G8B8toA8B8G8R8};
 
-		if (FeatureAvailable[IRR_GL_EXT_texture_format_BGRA8888])
+		if (FeatureAvailable[IRR_GL_EXT_texture_format_BGRA8888] || FeatureAvailable[IRR_GL_APPLE_texture_format_BGRA8888])
 			TextureFormats[ECF_A8R8G8B8] = {GL_BGRA, GL_BGRA, GL_UNSIGNED_BYTE};
-		else if (FeatureAvailable[IRR_GL_APPLE_texture_format_BGRA8888])
-			TextureFormats[ECF_A8R8G8B8] = {BGRA8_EXT, GL_BGRA, GL_UNSIGNED_BYTE};
```
Comment (grorp) [inline irr/src/OpenGLES2/Driver.cpp:92]: comparing https://registry.khronos.org/OpenGL/extensions/EXT/EXT_texture_format_BGRA8888.txt and https://registry.khronos.org/OpenGL/extensions/APPLE/APPLE_texture_format_BGRA8888.txt, looks like the APPLE version (contrary to the EXT version) has always defined `GL_BGRA8_EXT`, so we can rely on it

why did you make this change, making this code path also rely on the workaround?

**Comment by grorp** [comment]
@sfan5 crash on startup I noticed while testing this on my phone btw, probably related to your recent mainmenu sandboxing PR

```
ERROR[Main]: GUIEngine: execution of menu script failed: Failed to load and run script from /storage/emulated/0/Android/data/net.minetest.minetest/files/Minetest/builtin/init.lua:
ERROR[Main]: "Mod security: Blocked attempted write to /data/user/0/net.minetest.minetest/cache/cdb"
ERROR[Main]: stack traceback:
ERROR[Main]: 	[C]: in function 'create_dir'
ERROR[Main]: 	...es/Minetest/builtin/mainmenu/content/update_detector.lua:29: in main chunk
ERROR[Main]: 	[C]: in function 'dofile'
ERROR[Main]: 	...inetest/files/Minetest/builtin/mainmenu/content/init.lua:22: in main chunk
ERROR[Main]: 	[C]: in function 'dofile'
ERROR[Main]: 	...netest.minetest/files/Minetest/builtin/mainmenu/init.lua:47: in main chunk
ERROR[Main]: 	[C]: in function 'dofile'
ERROR[Main]: 	...ta/net.minetest.minetest/files/Minetest/builtin/init.lua:70: in main chunk
ERROR[Main]: No future without main menu!
Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 32750 (Main), pid 32403 (SDLActivity)
Process name is net.minetest.minetest, not key_process
Cmdline: net.minetest.minetest
pid: 32403, tid: 32750, name: Main  >>> net.minetest.minetest <<<
      #01 pc 000000000074cabc  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (GUIEngine::GUIEngine(JoystickController*, irr::gui::IGUIElement*, RenderingEngine*, IMenuManager*, MainMenuData*, bool&)+2228) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #02 pc 0000000000675b98  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (ClientLauncher::main_menu(MainMenuData*)+872) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #03 pc 0000000000674a54  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (ClientLauncher::launch_game(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >&, bool, GameStartData&, Settings const&)+1328) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #04 pc 0000000000672a2c  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (ClientLauncher::run(GameStartData&, Settings const&)+2108) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #05 pc 00000000008dd58c  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (main+8524) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #06 pc 00000000009126d0  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (SDL_Main+104) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #07 pc 0000000000e52470  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/lib/arm64/libluanti.so (Java_org_libsdl_app_SDLActivity_nativeRunMain+708) (BuildId: b6f4854c0fb8a5ac8136c1ef377f3f148f410c84)
      #10 pc 0000000000299388  /data/app/~~7-_rbBi9lWt-uJrpqbk7ng==/net.minetest.minetest-bgV-uzL2-i6lmzcQkrU7RQ==/oat/arm64/base.vdex (org.libsdl.app.SDLMain.run+152)
```

File: src/defaultsettings.cpp (hunk 1 above)
Comment (sfan5) [reply src/defaultsettings.cpp:569]: That's not the plan. I was only considering enabling it by default.

> the last time I tested, "enable_post_processing = false" still resulted in a performance improvement on my laptop

I hoping that `post_processing_texture_bits` could mend that to some degree

File: irr/include/SColor.h [hunk 3]
```diff
@@ -72,6 +72,9 @@ enum ECOLOR_FORMAT
 	//! 32 bit format using 16 bits for the red and green channels.
 	ECF_R16G16,
 
+	//! 32 bit format using 10 bits for R, G, B and 2 for alpha.
+	ECF_A2R10G10B10,
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/include/SColor.h:76]: i knew we needed a static_assert

File: irr/src/COpenGLCoreTexture.h [hunk 4]
```diff
@@ -170,6 +170,13 @@ class COpenGLCoreTexture : public ITexture
 			return;
 		}
 
+		// Problem: on some drivers ECF_A8R8G8B8 is mapped to GL_BGRA,
+		// which is an unsized format. Render targets however need to be created
+		// with a sized format. Since we don't upload texture data to RTTs, we can
+		// safely choose another comparable format.
+		if (InternalFormat == GL_BGRA)
+			InternalFormat = GL_RGBA8;
```
[the remark this answers is not in the document]
Comment (sfan5) [reply irr/src/COpenGLCoreTexture.h:176]: I got some things mixed up and this comment is incorrect.
please test the new commit

**Comment by AbduSharif** [comment]
How is quality/precision ate now with this PR?

**Comment by sfan5** [comment]
@grorp ping

> Is debanding still good enough with this?

10-bit is plenty.

File: irr/src/COpenGLCoreTexture.h [hunk 5]
```diff
@@ -170,6 +170,19 @@ class COpenGLCoreTexture : public ITexture
 			return;
 		}
 
+#ifndef IRR_COMPILE_GL_COMMON
+		// TODO explain this correctly
+		if (InternalFormat == GL_BGRA && Driver->Version.Major >= 3) {
+			InternalFormat = GL_RGBA8;
+			PixelFormat = GL_RGBA;
+		}
```
Comment (grorp) [inline irr/src/COpenGLCoreTexture.h:178]: `post_processing_texture_bits = 8` works now on my device with OpenGL ES 3.2

File: irr/src/COpenGLCoreTexture.h [hunk 6]
```diff
@@ -170,6 +170,19 @@ class COpenGLCoreTexture : public ITexture
...
+		// TODO is this needed?
+		if (InternalFormat == GL_BGRA && Driver->Version.Major < 3) {
+			InternalFormat = GL_RGBA;
+			PixelFormat = GL_RGBA; 
+		}
```
Comment (grorp) [inline irr/src/COpenGLCoreTexture.h:183]: I can't test this one

File: src/defaultsettings.cpp [hunk 7]
```diff
@@ -565,7 +566,10 @@ void set_default_settings()
 	settings->setDefault("active_block_range", "2");
 	settings->setDefault("viewing_range", "50");
 	settings->setDefault("leaves_style", "simple");
+	// Note: OpenGL ES 2.0 is not guaranteed to provide depth textures,
```
Comment (grorp) [inline src/defaultsettings.cpp:569]: well, OpenGL ES 2.0 is even guaranteed not to provide depth textures, strictly speaking

**Review by grorp — COMMENTED** [review]
Patch used to make it testable on Android: grorp/minetest@a7bfc3c7f4b791c5af0cab22a328033bc0592933

Note: I still need to check this PR for correctness. Doing that only makes sense once you've resolved your TODOs.

File: irr/src/COpenGLCoreTexture.h (hunk 5 above)
Comment (sfan5) [reply irr/src/COpenGLCoreTexture.h:178]: what's the performance difference between 8/10/16?

**Comment by AbduSharif** [comment]
With the patch I was able to open the game, but I'm seeing this with both and 10, 16, 8 bit:

```
ERROR[Main]: addPostProcessing(): color = A2R10G10B10 depth = D24
ERROR[Main]: addPostProcessing(): color = A16B16G16R16F depth = D24
ERROR[Main]: addPostProcessing(): color = A8R8G8B8 depth = D24
ERROR[Main]: Irrlicht: TODO explain this correctly triggered
```

Did the feature feature work?

**Comment by AbduSharif** [comment]
Here are some test screenshots, Repixture with Tonemapping:

Release, Tonemapped:
release_tonemapped

Release, Tonemapped and Debanded:
release_tonemapped_debanded

Comparisons, Tonemapping and Debanding are enabled:
PR, 8 bit:
pr_tonemapped_debanded_8bit

PR, 10:
pr_tonemapped_debanded_10bit

PR, 16:
pr_tonemapped_debanded_16bit

**Comment by AbduSharif** [comment]
https://imgsli.com/MzIwOTEz
Imgsli link if anyone wanted to scrutinize between 10bit and 16bit.

**Comment by sfan5** [comment]
While the difference with 8-bit is obvious, the JPEG compression of your screenshots makes reliably telling whether 10-bit or 16-bit is any different impossible.

My takeaways:
* we should automatically disable dithering if we only have an 8-bit surface, since it's pointless
* preferring 10-bit over 16-bit is safe (and _potentially_ good for performance)

**Comment by AbduSharif** [comment]
Here's another one with a device that takes screenshots as PNGs:

Release, Tonemapped:
release_tonemapped

10bit vs 16bit:
https://imgsli.com/MzIxMzM5

**Comment by sfan5** [comment]
> 10bit vs 16bit:
> https://imgsli.com/MzIxMzM5

still jpeg

**Comment by sfan5** [comment]
> Unless imgsli did something, its not.

even better: they're serving it as webp with incorrect content-type
grafik

**Comment by AbduSharif** [comment]
Volumetric lighting works on Xiaomi Redmi 9 (Mali-G52 MC2), but the quality is significantly reduced:

16bit:
16bit

10bit:
10bit

**Comment by lhofhansl** [comment]
10bit does something to bloom, I had noticed that too, but wanted to verify before I post it.

**Comment by sfan5** [comment]
we still don't have info on performance differences with different post_processing_texture_bits (I am a bit short on time to test myself) but it sounds like it's worth keeping as an advanced setting at least.

**Comment by lhofhansl** [comment]
With my NVidia card on my Laptop, I get a *very* modest gain (44 FPS to 45 FPS) with 10 texture bits.

File: builtin/settingtypes.txt [hunk 8]
```diff
@@ -468,9 +468,39 @@ smooth_lighting (Smooth lighting) bool true
...
+#
+#    Requires: enable_dynamic_shadows, opengl
+shadow_update_frames (Map shadows update frames) int 8 1 16
+
+#    Set the soft shadow radius size.
+#    Lower values mean sharper shadows, bigger values mean softer shadows.
+#    Minimum value: 1.0; maximum value: 15.0
+#
+#    Requires: enable_dynamic_shadows, opengl
+shadow_soft_radius (Soft shadow radius) float 5.0 1.0 15.0
```
Comment (grorp) [inline builtin/settingtypes.txt:502]: advanced settings need to be put into a section in the "Advanced" category so they're actually less visible / hidden when advanced settings are hidden

Comment (sfan5) [reply builtin/settingtypes.txt:502]: settingtypes.txt says:
```
# There shouldn't be too much settings per category; settings that shouldn't be
#  modified by the "average user" should be in (sub-)categories called "Advanced".
```

if this is incorrect then we need to fix the documentation.

Comment (grorp) [reply builtin/settingtypes.txt:502]: Yeah, that comment is wrong. The current advanced settings structure was introduced in #12490 and implemented in #13861. The "Client and Server", "Mapgen" and "Advanced" top-level categories are the ones containing advanced settings.

Comment (sfan5) [reply builtin/settingtypes.txt:502]: then this category doesn't work either I guess
https://github.com/minetest/minetest/blob/df4e70b2c7631317430bd78061507270d23db67b/builtin/settingtypes.txt#L1821-L1829

Comment (grorp) [reply builtin/settingtypes.txt:502]: That's the "Advanced" category in the "Advanced" top-level category, so it works.

(By "top-level category", I mean the headings without asterisks (-> thus having the highest level in the category hierarchy). These are shown as yellow labels in the GUI.)

Comment (sfan5) [reply builtin/settingtypes.txt]: I discovered further unclear behavior of the advanced toggle but the settings are now moved.

File: builtin/settingtypes.txt [hunk 9]
```diff
@@ -77,8 +77,11 @@
 # Sections are marked by a single line in the format: [Section Name]
 # Sub-section are marked by adding * in front of the section name: [*Sub-section]
 # Sub-sub-sections have two * etc.
-# There shouldn't be too much settings per category; settings that shouldn't be
-#  modified by the "average user" should be in (sub-)categories called "Advanced".
+# There shouldn't be too many settings per category.
+#
+# The top-level category "Advanced" is handled specially and its contents only
+# shown when a checkbox is checked. It contains settings not for the "average user".
+# FIXME: some other categories also depend on show_advanced and I don't know why.
```
Comment (grorp) [inline builtin/settingtypes.txt:84]: This FIXME shouldn't b here. These other categories depending on show_advanced is intended, they should just be documented too.

**Comment by HybridDog** [comment]
The differences of the banding with 10 and 16 bits are visible if I edit the screenshots with the GIMP. There is slightly more banding with 10 bits, which isn't noticeable in the unmodified screenshot for me and still looks better than 8 bit without dithering.
### 10 bit

10 close
### 16 bit

16 close

**Comment by HybridDog** [comment]
> Volumetric lighting breaks with 10 bit

Perhaps with 10 bit the colour values are constrained in [0, 1] whereas with 16 bit they can take any floating point number.

**Comment by HybridDog** [comment]
> We shouldn't be using this color format for intermediate results however.

I think values constrained in [0,1] make it difficult to add support for wide-gamut and/or very bright (HDR) displays and using integers instead of floats for the values makes it difficult to add support for gamma-correct rendering.

File: builtin/settingtypes.txt [hunk 10]
```diff
@@ -553,21 +551,6 @@ shadow_filters (Shadow filter quality) enum 1 0,1,2
...
-#
-#    Requires: enable_dynamic_shadows, opengl
-shadow_update_frames (Map shadows update frames) int 8 1 16
-
-#    Set the soft shadow radius size.
-#    Lower values mean sharper shadows, bigger values mean softer shadows.
-#    Minimum value: 1.0; maximum value: 15.0
-#
-#    Requires: enable_dynamic_shadows, opengl
-shadow_soft_radius (Soft shadow radius) float 5.0 1.0 15.0
```
Comment (lhofhansl) [inline builtin/settingtypes.txt:569]: Soft radius does not strike me as an advanced setting.

Comment (sfan5) [reply builtin/settingtypes.txt:569]: moved it back.

**Comment by lhofhansl** [comment]
Works for me. Code looks reasonable :)

(I do not see any difference between 8, 10, 16 bits - OpenGL 4.6.0 + SDL.)

**Pull request merged**